File: system_simulations/vending_machine/cpp/discrete_time/VendingMachine.cpp
#include <iostream>
#include <string>
#include "VendingMachine.h"

using namespace std;

VendingMachine::VendingMachine(int init_quarters, int init_nickels, int init_dimes){
	this->quarters = init_quarters;
	this->nickels = init_nickels;
	this->dimes = init_dimes;
	this->value = 0;
	this->cancel = false;
}

int* VendingMachine::lambda(){
	int coffee=0, nothing=0;

	if (cancel){
		int* change = dispense_change();
		int* y = new int[5]{coffee, change[0], change[1], change[2], nothing};
		return y;
	} else if (value >= 100){
		coffee = value / 100;
	}

	if (!coffee){
		nothing = 1;
	}

	int* y = new int[5]{coffee, 0, 0, 0, nothing};
	return y;
}

void VendingMachine::delta(int argc, char* args){
	if (cancel){
		cancel = false;
		if (value > 0){
			change_state();
		}
	} else if (value >= 100){
		value %= 100;
	}

	for (int i = 0; i < argc; i++){
		switch (args[i]){
			case 'q': quarters++; value += 25; break;
			case 'n': nickels++;  value += 5;  break;
			case 'd': dimes++; 	  value += 10; break;
			case 'c': cancel=true; break;
			case 'w': return;
			default: cout << "Invalid input"; break;
		}
	}
}

void VendingMachine::print_state(){
	cout << "Quarters: " << quarters << " || Nickels: " << nickels << 
	" || Dimes: " << dimes << " || Value: " << value << " || Cancel: " << cancel << endl;
}

void VendingMachine::change_state(){
	int* change = dispense_change();
	for (int i = 0; i < change[0]; i++){
		value -= 25;
		quarters--;
	}

	for (int i = 0; i < change[1]; i++){
		value -= 5;
		nickels--;
	}

	for (int i = 0; i < change[2]; i++){
		value -= 10;
		dimes--;
	}
}

int* VendingMachine::dispense_change(){
	int q_count=0, n_count=0, d_count=0;
	int tmp_value = value;
	while (tmp_value != 0){
		while (tmp_value >= 25){
			if (quarters > 0){
				if (tmp_value < 50 && tmp_value % 10 == 0){
					if (dimes > 0){
						break;
					}
				}
				q_count++;
				tmp_value -= 25;
			} else {
				break;
			}
		}
		while (tmp_value >= 10){
			if (dimes > 0){
				d_count++;
				tmp_value -= 10;
			} else {
				break;
			}
		}
		while (tmp_value >= 5){
			if (nickels > 0){
				n_count++;
				tmp_value -= 5;
			} else {
				break;
			}
		}
		if (tmp_value != 0){
			cout << "OUT OF SUFFICIENT CHANGE" << endl;
			break;
		}
	}

	int* pointer = new int[3]{q_count, n_count, d_count};
	return pointer;
}

File: system_simulations/vending_machine/java/discrete_time/VendingMachine.java
public class VendingMachine{
	private int quarters=0, nickels=0, dimes=0, value=0;
	private boolean cancel = false;
	private int state = 0;

	public VendingMachine(int init_quarters, int init_nickels, int init_dimes){
		this.quarters = init_quarters;
		this.nickels = init_nickels;
		this.dimes = init_dimes;	
	}
	
	public void print_state(){
		System.out.println("State #" + state + ":\nquarters: " + quarters + " || nickels: " + nickels
		+ " || dimes: " + dimes + " || value: " + value + " || cancel: " + cancel + "\n");
	}

	// lambda and delta cannot depend on each other (can call similar code in both lambda and delta)
	public int[] lambda(){
		int coffee=0, nothing=0;

		if (cancel){
			try{
				int[] change = dispense_change();
				int[] y = {coffee, change[0], change[1], change[2], nothing};
				return y;
			} catch (AtomicModelException e){
				e.printStackTrace();
				System.out.println("OUT OF ORDER");
				System.exit(0);
			}
		} else if (value >= 100){
			coffee = value/100;
		}

		if (coffee == 0){
			nothing = 1;
		}

		int[] y = {coffee, 0, 0, 0, nothing};
		return y;
	}

	public void delta(String[] args){
		state++;
		if (cancel){
			cancel = false;
			if (value > 0){
				changeState();
			}
		} else if (value >= 100){
			value %= 100;
		}

		for (int i = 0; i < args.length; i++){
			switch (args[i]){
				case "q": quarters++; value += 25;break;
				case "n": nickels++; value += 5;break;
				case "d": dimes++; value += 10; break;
				case "c": cancel = true; break;
				case " ": break;
				default : System.out.println("Something went wrong");
			}
		}
	}

	private void changeState(){
		try {
			int[] change = dispense_change();
			for (int i = 0; i < change[0]; i++){
				value -= 25;
				quarters--;
			}
			for (int i = 0; i < change[1]; i++){
				value -= 5;
				nickels--;
			}
			for (int i = 0; i < change[2]; i++){
				value -= 10;
				dimes--;
			}
	} catch (AtomicModelException e){
			e.printStackTrace();
			System.out.println("\nOUT OF ORDER...");
			System.exit(0);		
		}
	}

	private int[] dispense_change() throws AtomicModelException{
		int q_count=0, n_count=0, d_count=0;
		if (value != 0){
			while(value != 0){
				while (value >= 25){
					if (quarters > 0){
						if (value < 50 && value % 10 == 0){
							if (dimes > 0){
								break;
							}
						}
						value -= 25;
						quarters--;
						q_count++;
					} else {
						break;
					}
				}

				while (value >= 10){
					if (dimes > 0){
						value -= 10;
						dimes--;
						d_count++;
					} else {
						break;
					}
				}

				while (value >= 5){
					if (nickels > 0){
						value -= 5;
						nickels--;
						n_count++;
					} else {
						break;
					}
				}
				if (value != 0){
					throw new AtomicModelException("Insufficent Change. Please call xxx-xxx-xxxx for more information");
				}
			}
		}
		int[] change = {q_count, n_count, d_count};
		return change;
	}
}
